File: libs/east-diagnostics/src/rules/no-untracked-east-data.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type * as ts from "typescript";
import type { EastRule, RuleContext } from "../types.js";
import { insideBlockScope } from "../block-scope.js";
import { isEastExprType } from "../east-type.js";

const NAME = "no-untracked-east-data";
const CODE = 990013;

// Inside an East block, data destined for an East-typed position (a factory
// `data` argument, a JSX `data={...}` prop — anything whose contextual type
// accepts an `Expr`) must be bound with `$.const` / `$.let`, not a bare JS
// `const` literal. A bare `const` isn't tracked by the block builder: it
// carries no East type at the binding, and consuming it twice inlines the
// literal twice in the IR. TS-side config (column defs, encodings, `days`
// string lists) has a non-Expr contextual type and is exempt naturally.

/** The plain-JS-literal initializer of `decl`, or undefined. */
function plainLiteralInitializer(
  decl: ts.VariableDeclaration,
  t: RuleContext["ts"],
): ts.Expression | undefined {
  const init = decl.initializer;
  if (init === undefined) return undefined;
  const unwrapped = t.isAsExpression(init) || t.isSatisfiesExpression(init) ? init.expression : init;
  if (t.isArrayLiteralExpression(unwrapped) || t.isObjectLiteralExpression(unwrapped)) {
    return unwrapped;
  }
  return undefined;
}

export const noUntrackedEastData: EastRule = {
  name: NAME,
  code: CODE,
  description:
    "Inside East blocks, bind data consumed in East-typed positions with $.const/$.let, not a bare JS const.",
  check(node, ctx) {
    const t = ctx.ts;
    if (!t.isIdentifier(node)) return;

    // Only identifiers in a consuming position: a call argument or a JSX
    // attribute expression value.
    const parent = node.parent;
    const isCallArg =
      parent !== undefined &&
      t.isCallExpression(parent) &&
      parent.expression !== node &&
      parent.arguments.some((arg) => arg === node);
    const isJsxValue = parent !== undefined && t.isJsxExpression(parent) && parent.expression === node;
    if (!isCallArg && !isJsxValue) return;
    if (!insideBlockScope(node, ctx)) return;

    // The position must be East-typed (its contextual type accepts an Expr).
    const contextual = ctx.checker.getContextualType(node);
    if (contextual === undefined || !isEastExprType(contextual)) return;

    // The identifier must resolve to a bare-const plain literal declared
    // inside an East block scope.
    const symbol = ctx.checker.getSymbolAtLocation(node);
    const decl = symbol?.valueDeclaration;
    if (decl === undefined || !t.isVariableDeclaration(decl)) return;
    if (plainLiteralInitializer(decl, t) === undefined) return;
    if (!insideBlockScope(decl, ctx)) return;

    const sf = ctx.sourceFile;
    const start = node.getStart(sf);
    ctx.report({
      ruleName: NAME,
      code: CODE,
      start,
      length: node.getEnd() - start,
      messageText:
        `Bare \`const ${node.text} = …\` isn't tracked by the East block builder. Bind East data with ` +
        "`$.const([...], Type)` (or `$.let`) so the binding carries its East type and is evaluated once.",
      category: "suggestion",
    });
  },
};

File: libs/east-py/packages/east-py/east/diagnostics/rules/no_untracked_east_data.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``no-untracked-east-data``: data a body feeds to an East expression must be
bound with ``b.let`` / ``b.const``, not held in a plain python local.

A plain local carries no East type at the binding, and the builder re-inlines
its literal at every use — two uses, two copies in the IR. The block builder
binds it once and gives it a type. The TypeScript rule of the same name, which
reads the contextual type; here the tell is a python literal local reaching an
expression's method.
"""

from __future__ import annotations

import ast

from east.diagnostics.types import Body, Context, body_nodes


def message(name: str) -> str:
    return (f"`{name}` is a plain python local, so it carries no East type and is re-inlined at "
            f"every use — bind it with b.const({name}, Type) (or b.let) and pass the binding")


class NoUntrackedEastData:
    name = "no-untracked-east-data"
    code = 14
    category = "suggestion"
    supersedes: tuple[str, ...] = ()
    description = ("Data reaching an East expression must be bound with b.const / b.let, not held "
                   "in a plain python local.")

    def check(self, body: Body, ctx: Context) -> None:
        untracked = _literal_locals(body)
        if not untracked:
            return
        for node in body_nodes(body):
            if not (isinstance(node, ast.Call) and isinstance(node.func, ast.Attribute)
                    and body.is_expression(node.func.value)):
                continue
            for arg in [*node.args, *(k.value for k in node.keywords)]:
                if isinstance(arg, ast.Name) and arg.id in untracked:
                    ctx.report(arg, self, message(arg.id))


def _literal_locals(body: Body) -> set[str]:
    """Names the body binds to a bare python list/dict literal — an EMPTY one
    is `prefer-explicit-east-type`'s to talk about, and a name the block
    builder bound is tracked already."""
    names: set[str] = set()
    for node in body_nodes(body):
        if not (isinstance(node, ast.Assign) and len(node.targets) == 1
                and isinstance(node.targets[0], ast.Name)):
            continue
        target = node.targets[0].id
        if target in body.expr_names:
            continue
        value = node.value
        if (isinstance(value, ast.List) and value.elts) or (isinstance(value, ast.Dict) and value.keys):
            names.add(target)
    return names
